### PR TITLE
Add "migrate down" command to roll back to last major version schema

### PR DESCRIPTION
**Pull request by john-metabase**
Background: when updating Metabase to a new version, there is no way to downgrade the app DB to a previous version without a database backup.

This change introduces a new `migrate down` command, and tests to ensure that all migrations added from now on can be rolled back to the last schema version of v44 and back up again.

The migration YAML linter has also been updated to check for a `rollback` key where any Liquibase change type doesn't support auto rollback for all migrations added from v45 onwards.

Pseudo-terminal session example:

```
# Run v44 to create an app db, optionally set it up and create users, dashboards, etc.
$ java -jar metabase-0.44.0.jar
...snip...
2022-11-11 18:51:13,028 INFO metabase.core :: Metabase Initialization COMPLETE
^C

# Run v45 to migrate app db up (or use `migrate up` to skip starting the app if you're in a hurry):
$ clojure -M:run
...snip...
^C

# Use the `migrate down` command from v45 (local branch) to roll back the schema:
$ clojure -M:run migrate down
...snip...
2022-11-11 18:45:33,907 INFO db.setup :: Setting up Liquibase...
2022-11-11 18:45:34,218 INFO db.setup :: Liquibase is ready.
2022-11-11 18:45:34,223 INFO db.liquibase :: Rolling back app database schema to version 44


# Verify that v44 starts with the rolled-back app DB:
$ java -jar metabase-0.44.0.jar
```

The app DB will be rolled back to v44 and will work with the original v44 JAR which created it.

### CODE REVIEW

File: resources/migrations/000_migrations.yaml
Comment (camsaul) [inline resources/migrations/000_migrations.yaml:12897]: I don't think we really need to roll these back -- changing from `text` to `longtext` really shouldn't make any difference other than `longtext` having a larger max size

File: bin/lint-migrations-file/src/change_set/strict.clj
Comment (camsaul) [inline bin/lint-migrations-file/src/change_set/strict.clj:74]: I would rather error on the side of whitelisting the ones that have auto-rollback as opposed to blacklisting ones that do not... maybe at some point in the future Liquibase adds a new change type that doesn't support auto-rollback and we use it and break reverse migrations as a result. Might be far-fetched but the linter is supposed to save us from doing something dumb so I'd rather error on the side of caution

File: bin/lint-migrations-file/src/change_set/strict.clj
Comment (camsaul) [inline bin/lint-migrations-file/src/change_set/strict.clj:96]: It might not be a bad idea to add some tests around this. Just to be sure we're actually catching things we think we're catching

File: src/metabase/db/liquibase.clj
Comment (camsaul) [inline src/metabase/db/liquibase.clj:223]: It might not be the best idea in the whole universe to order by when things were executed... it's totally possible that things get executed out of order, for example if I am running some branch locally where I add `v45.00-100` and then I switch back to master and it has `v45.00-099` which was missing from my branch... in that case `-100` would have been executed before `-099`. 

Probably doesn't matter in that case but we do want to leave the option open to add new migrations to minor version releases of previous major version releases, which is why our ID format is `v<major>.<minor>-<migration>` as opposed to `v<major>-<migration>`. We might add a `v44.01-001` for some reason or another -- we've done it once in the past and we might have to do it again for a bug or security fix.

`ORDER BY id` should work tho.

File: src/metabase/db/liquibase.clj
Comment (camsaul) [inline src/metabase/db/liquibase.clj:218]: if we run this with `java -jar` are there any situations where `target-version` can come in as a string and need to be parsed?

File: src/metabase/db/liquibase.clj
Comment (camsaul) [inline src/metabase/db/liquibase.clj:223]: I'm surprised this works on MySQL without quoting identifiers since for "reasons" we have the table named as `DATABASCHANGELOG` (all caps) and MySQL lowercases unquoted identifiers by default. Are the tests passing for MySQL? Maybe our tests are busted if they are.

**Review by camsaul — APPROVED** [review]
From my end this looks good overall, but we should try to figure out how MySQL tests managed to pass even tho that SQL looks like it shouldn't work (if they did pass?). Also we should manually check that this works with the JAR if Clojure CLI arg parsing isn't applied... 

You should get a second review from @flamber before merging since I know he's going to be using this a lot, we want to make sure he gets a chance to try it out and give it a 👍 .

Oh yeah, don't forget to mark this `Backport` so it gets copied over to the 45 release branch

File: resources/migrations/000_migrations.yaml
Comment (john-metabase) [reply resources/migrations/000_migrations.yaml:12897]: Good point.

File: bin/lint-migrations-file/src/change_set/strict.clj
Comment (john-metabase) [reply bin/lint-migrations-file/src/change_set/strict.clj:74]: That's an excellent point, thanks. Will adjust.

File: src/metabase/db/liquibase.clj
Comment (john-metabase) [reply src/metabase/db/liquibase.clj:223]: Unfortunately Liquibase does the rollback strictly sequentially by order applied, and not by ID, so we have to order by `orderexecuted` to get the same sequence Liquibase will see when it rolls back. You're right that could potentially cause some weird issues in a dev environment, but only if someone is using the rollback command, so it should not be an issue.

File: src/metabase/db/liquibase.clj
Comment (john-metabase) [reply src/metabase/db/liquibase.clj:218]: No, the target version is not an argument to the command at this point.

File: src/metabase/db/liquibase.clj
Comment (john-metabase) [reply src/metabase/db/liquibase.clj:223]: I've tested with MariaDB, and the rollback test succeeds on all drivers in CI.

File: bin/lint-migrations-file/src/change_set/strict.clj
Comment (john-metabase) [reply bin/lint-migrations-file/src/change_set/strict.clj:96]: Done.

**Comment by john-metabase** [comment]
Still looking for review/approval from @flamber 🙏 on this one before merging. I just pushed a tiny commit to remove the `trs` calls.

**Review by flamber — APPROVED** [review]

**Pull request merged**
